apps/web/app/.server/service/signatures.ts

```diff
@@ -0,0 +1,315 @@
+import { db } from "@/.server/db";
+import {
+  getEmergencyProposalByExternalId,
+  updateEmergencyProposal,
+} from "@/.server/db/dto/emergency-proposals";
+import {
+  createOrIgnoreSignature,
+  getSignaturesByEmergencyProposalId,
+} from "@/.server/db/dto/signatures";
+import { getUserAuthRole } from "@/.server/service/authorized-users";
+import { getFreezeProposalSignatureArgs } from "@/common/freeze-proposal";
+import { getL2CancellationSignatureArgs } from "@/common/l2-cancellations";
+import { emergencyProposalStatusSchema } from "@/common/emergency-proposal-status";
+import { signActionEnum } from "@/common/sign-action";
+import { GUARDIANS_COUNCIL_THRESHOLD, SEC_COUNCIL_THRESHOLD } from "@/utils/emergency-proposals";
+import { badRequest, notFound } from "@/utils/http";
+import { type BasicSignature, classifySignatures } from "@/utils/signatures";
+import { type Hex, hexToBigInt } from "viem";
+import {
+  assertSignatureIsValidMultisig,
+  assertValidSignatureZkFoundation,
+} from "@/.server/service/verify-signature-multisig";
+import {
+  emergencyUpgradeActionForRole,
+  regularUpgradeContractNameByRole,
+  standardUpgradeActionForRole,
+  UserRoleEnum,
+} from "@/common/user-role-schema";
+import { getProposalByExternalId } from "@/.server/db/dto/proposals";
+import { getFreezeProposalById } from "@/.server/db/dto/freeze-proposals";
+import { getL2CancellationById } from "@/.server/db/dto/l2-cancellations";
+import { freezeActionFromType } from "@/common/freeze-proposal-type";
+import { multisigContractForRole } from "@/.server/user-role-data";
+import { guardianMembers } from "./ethereum-l1/contracts/guardians";
+import { securityCouncilMembers } from "./ethereum-l1/contracts/security-council";
+import { zkFoundationAddress } from "./ethereum-l1/contracts/emergency-upgrade-board";
+import {
+  emergencyUpgradeBoardAddress,
+  guardiansAddress,
+  securityCouncilAddress,
+} from "./ethereum-l1/contracts/protocol-upgrade-handler";
+
+async function shouldMarkProposalAsReady(allSignatures: BasicSignature[]): Promise<boolean> {
```

Comment:
This seems to be emergency upgrade specific. Let's reflect it in the function name.